### PR TITLE
refactor: extract checkpointing logic into standalone Checkpointer class

### PR DESCRIPTION
**Pull request by prasannakotyal**
- Create new Checkpointer class to handle checkpoint saving/loading
- Extract _get_missing_data, _save_intermediate_checkpoint methods from SDG
- Improve separation of concerns and testability
- Maintain existing functionality while reducing coupling
- All tests are passing (32/32). I've also fixed some minor style issues.

Note: I couldn't run the action and markdown linters locally on Windows, but CI will catch any issues.

Fixes: Tight coupling between data generation and checkpointing logic"

## Summary by CodeRabbit

- **New Features**
  - Introduced checkpoint management for SDG data generation workflows, allowing progress to be saved and resumed.
  - Added a conditional prompt block enabling dynamic prompt selection based on input data.
  - Added iterative block functionality to generate multiple outputs by repeated processing.
  - Introduced a dataset filtering block to filter samples based on column values with flexible operations.
- **Refactor**
  - Streamlined checkpointing and missing data logic for improved maintainability and clarity.
  - Replaced wildcard imports with explicit exports for better package clarity.
  - Enhanced documentation and type annotations across multiple blocks for improved usability.
  - Refactored skill classification configuration with detailed categories, principles, and examples.
- **Tests**
  - Added comprehensive tests for checkpointing functionality to ensure reliability.
  - Added new unit tests covering dataset filtering block behavior and edge cases.
- **Chores**
  - Updated configuration to ignore the `sdg_env/` directory in version control.
  - Updated model configurations in skill generation flows to use newer models.
  - Upgraded Flask-WTF dependency version in the web interface requirements.

### CODE REVIEW

File: src/sdg_hub/checkpointer.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        Args:
+            dataset: Dataset to save as checkpoint
+        """
+        if self.checkpoint_dir is None:
+            return
+            
+        checkpoint_id = uuid.uuid4().hex
+        checkpoint_file = f"{self.checkpoint_dir}/data_checkpoint_{checkpoint_id}.jsonl"
+        logger.info(f"Saving checkpoint to {checkpoint_file}")
+        dataset.to_json(checkpoint_file, orient="records", lines=True)
```
Comment (shivchander) [inline src/sdg_hub/checkpointer.py:125]: could we use the `safe_concatenate_datasets` here?

**Review by shivchander — APPROVED** [review]
Thanks for the contribution!
